tree/ntupleutil/v7/src/RNTupleInspector.cxx

```diff
@@ -22,111 +22,219 @@
-   std::uint64_t compressedSize = 0;
-   std::uint64_t uncompressedSize = 0;
+   fDescriptor = descriptorGuard->Clone();
+}
+
+void ROOT::Experimental::RNTupleInspector::CollectColumnInfo()
+{
+   fOnDiskSize = 0;
+   fInMemorySize = 0;
+
+   for (DescriptorId_t colId = 0; colId < fDescriptor->GetNPhysicalColumns(); ++colId) {
+      const RColumnDescriptor &colDesc = fDescriptor->GetColumnDescriptor(colId);
+
+      // We generate the default memory representation for the given column type in order
+      // to report the size _in memory_ of column elements.
+      auto colType = colDesc.GetModel().GetType();
+      std::uint32_t elemSize = ROOT::Experimental::Detail::RColumnElementBase::Generate(colType)->GetSize();
+      std::uint64_t nElems = 0;
+      std::uint64_t onDiskSize = 0;
+      std::uint64_t inMemSize = 0;
+
+      for (const auto &clusterDescriptor : fDescriptor->GetClusterIterable()) {
+         if (!clusterDescriptor.ContainsColumn(colId)) {
+            continue;
+         }
+
+         auto columnRange = clusterDescriptor.GetColumnRange(colId);
+         nElems += columnRange.fNElements;
+
+         if (fCompressionSettings == -1) {
+            fCompressionSettings = columnRange.fCompressionSettings;
+         } else {
+            R__ASSERT(columnRange.fCompressionSettings == fCompressionSettings);
+         }
+
+         const auto &pageRange = clusterDescriptor.GetPageRange(colId);
+
+         for (const auto &page : pageRange.fPageInfos) {
+            onDiskSize += page.fLocator.fBytesOnStorage;
+            inMemSize += page.fNElements * elemSize;
+            fOnDiskSize += page.fLocator.fBytesOnStorage;
+            fInMemorySize += page.fNElements * elemSize;
+         }
+      }
+
+      fColumnInfo.emplace(colId, RColumnInfo(colDesc, onDiskSize, inMemSize, elemSize, nElems));
+   }
+}
+
+void ROOT::Experimental::RNTupleInspector::CollectFieldInfo()
+{
+   std::deque<DescriptorId_t> fieldIdQueue{fDescriptor->GetFieldZeroId()};
 
-   for (const auto &clusterDescriptor : descriptorGuard->GetClusterIterable()) {
-      compressedSize += clusterDescriptor.GetBytesOnStorage();
+   while (!fieldIdQueue.empty()) {
+      auto currId = fieldIdQueue.front();
+      fieldIdQueue.pop_front();
 
-      if (compressionSettings == -1) {
-         compressionSettings = clusterDescriptor.GetColumnRange(0).fCompressionSettings;
+      for (const auto &fieldDescriptor : fDescriptor->GetFieldIterable(currId)) {
+         DescriptorId_t fieldId = fieldDescriptor.GetId();
+         std::uint64_t onDiskSize = 0;
+         std::uint64_t inMemSize = 0;
+
+         for (const auto colId : GetColumnsForFieldTree(fieldId)) {
+            auto colInfo = GetColumnInfo(colId);
+            onDiskSize += colInfo.GetOnDiskSize();
+            inMemSize += colInfo.GetInMemorySize();
+         }
+
+         fFieldInfo.emplace(fieldId, RFieldTreeInfo(fieldDescriptor, onDiskSize, inMemSize));
+         fieldIdQueue.push_back(fieldId);
```

Comment:
This has quadratic complexity in the number of fields. I think we should use recursion to make it linear. I.e. given that all sub fields are processed, we can add to the on disk size and the in memory size the columns that are directly attached to the mother field.